src/com/ms/silverking/cloud/dht/daemon/RingMapState.java

```diff
@@ -49,136 +49,133 @@
+  static final int peerStateCheckIntervalMillis = 1 * 60 * 1000;
+  // Delay convergence slightly in order to allow in-flight operations being sent to the old
+  // replica set only to complete.
+  private static final int convergenceDelayMillis = 10 * 1000;
+
+  private static final boolean verbosePeerStateTransition = true;
+  static final boolean verbosePeerStateCheck = true;
+  static final boolean debug = true;
+
+  static final Timer peerStateWatcherTimer = new SafeTimer("PeerStateWatcherTimer", true);
+
+  RingMapState(IPAndPort nodeID, DHTMetaUpdate dhtMetaUpdate, RingID ringID, StoragePolicyGroup storagePolicyGroup,
+      com.ms.silverking.cloud.toporing.meta.MetaClient ringMC, ExclusionSet exclusionSet,
+      RingMapStateListener ringMapStateListener, com.ms.silverking.cloud.dht.meta.MetaClient dhtMC,
+      boolean convergenceAlreadyComplete) {
+    //RingConfiguration   ringConfig;
+
+    this.nodeID = nodeID;
+    dhtConfigVersion = dhtMetaUpdate.getDHTConfig().getZKID();
+    ringConfig = dhtMetaUpdate.getNamedRingConfiguration().getRingConfiguration();
+    this.rawRingTree = dhtMetaUpdate.getRingTree();
+    this.ringIDAndVersionPair = new RingIDAndVersionPair(ringID, rawRingTree.getRingVersionPair());
+    this.ringID = ringID;
+    this.storagePolicyGroup = storagePolicyGroup;
+    this.ringMapStateListener = ringMapStateListener;
+    this.dhtMC = dhtMC;
+
+    //exclusionSetLock = new ReentrantLock();
+    //exclusionSetCV = exclusionSetLock.newCondition();
+
+    curInstanceExclusionSet = ExclusionSet.emptyExclusionSet(0);
+    curExclusionSet = ExclusionSet.emptyExclusionSet(0);
+
+    secondarySetsRef = new AtomicReference();
+    try {
+      ringTreeMinusExclusions = RingTreeBuilder.removeExcludedNodes(rawRingTree, exclusionSet);
+    } catch (InvalidRingException ire) {
+      throw new RuntimeException("Unexpected InvalidRingException", ire);
+    }
+    resolvedReplicaMapMinusExclusions = ringTreeMinusExclusions.getResolvedMap(ringConfig.getRingParentName(), null);
+    try {
+      exclusionWatcher = new ExclusionWatcher(storagePolicyGroup, ringConfig, ringMC.createCloudMC());
+    } catch (Exception e) {
+      throw new RuntimeException("Exception creating ExclusionWatcher", e);
+    }
+    try {
+      ringStateZK = new RingStateZK(dhtMetaUpdate.getMetaClient(), dhtMetaUpdate.getDHTConfig(),
+          ringIDAndVersionPair.getRingVersionPair());
+    } catch (Exception e) {
+      throw new RuntimeException("Exception creating ringStateZK", e);
+    }
+
+    cp = new ConvergencePoint(dhtConfigVersion, ringIDAndVersionPair, rawRingTree.getRingCreationTime());
+
+    if (!convergenceAlreadyComplete) {
+      try {
+        peerStateWatcher = new PeerStateWatcher(getResolvedReplicaMap().allReplicas(), dhtMetaUpdate.getMetaClient(),
+            dhtMetaUpdate.getDHTConfig(), ringIDAndVersionPair.getRingVersionPair(), this);
+      } catch (KeeperException ke) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)